File: src/lib/builds/finishNextSlot.ts
import type { FinishCategory, FinishGap, FinishGapStatus } from "@/lib/builds/finishGaps";

export type FinishWalkthroughStep = "overview" | "category" | "fill" | "done";

export type FinishPostMutationTarget = {
  step: FinishWalkthroughStep;
  fillSlot: string | null;
  /** Prefer focusing this category after mutation when returning to overview. */
  category: FinishCategory | null;
};

/** Map finish gap category to create-set-attach type. */
export function finishCategoryToSetType(
  category: FinishCategory,
): "armor" | "weapon" | "mod" {
  return category;
}

/**
 * First empty required slot in gap order (already filtered/ordered by evaluateFinishGaps).
 */
export function firstEmptyRequiredSlot(gap: Pick<FinishGap, "emptySlots">): string | null {
  return gap.emptySlots[0] ?? null;
}

export type ResolvePostMutationStepInput = {
  gap: FinishGap | null | undefined;
  /**
   * When true (default), needs_fill + live covering auto-opens fill for first empty slot.
   * Snapshot covering never auto-fills.
   */
  autoEnterFill?: boolean;
};

/**
 * After create/capture/fill refresh, decide next Finish walkthrough step for the active category gap.
 */
export function resolvePostMutationStep(
  input: ResolvePostMutationStepInput,
): FinishPostMutationTarget {
  const { gap, autoEnterFill = true } = input;
  if (!gap || gap.status === "satisfied") {
    return { step: "overview", fillSlot: null, category: null };
  }

  if (gap.status === "needs_fill") {
    if (gap.coveringMode === "snapshot") {
      return {
        step: "category",
        fillSlot: null,
        category: gap.category,
      };
    }
    if (gap.coveringMode === "live" && gap.coveringSetId) {
      const slot = firstEmptyRequiredSlot(gap);
      if (slot && autoEnterFill) {
        return {
          step: "fill",
          fillSlot: slot,
          category: gap.category,
        };
      }
      return {
        step: "category",
        fillSlot: null,
        category: gap.category,
      };
    }
    // needs_fill without covering — treat as category
    return { step: "category", fillSlot: null, category: gap.category };
  }

  // needs_set | capture_available
  return {
    step: "category",
    fillSlot: null,
    category: gap.category,
  };
}

/** Statuses that show one-tap create (and optional capture). */
export function showFinishCreateActions(status: FinishGapStatus): boolean {
  return status === "needs_set" || status === "capture_available";
}

File: flutter/packages/domain/lib/src/evaluators/finish_next_slot.dart
/// Pure next-slot / post-mutation walkthrough helpers for finish flow.
///
/// Mirrors TypeScript `src/lib/builds/finishNextSlot.ts`.
library;

import '../models/equipment.dart';
import 'finish_gaps.dart';

/// Finish walkthrough step (TS `FinishWalkthroughStep`).
enum FinishWalkthroughStep {
  overview('overview'),
  category('category'),
  fill('fill'),
  armorOptimize('armor_optimize'),
  done('done');

  const FinishWalkthroughStep(this.wireName);
  final String wireName;

  static FinishWalkthroughStep? tryParse(String wire) {
    for (final v in FinishWalkthroughStep.values) {
      if (v.wireName == wire) return v;
    }
    return null;
  }
}

/// Target step after a finish mutation (TS `FinishPostMutationTarget`).
class FinishPostMutationTarget {
  const FinishPostMutationTarget({
    required this.step,
    this.fillSlot,
    this.category,
  });

  final FinishWalkthroughStep step;
  final String? fillSlot;
  final FinishCategory? category;

  @override
  bool operator ==(Object other) {
    if (identical(this, other)) return true;
    return other is FinishPostMutationTarget &&
        other.step == step &&
        other.fillSlot == fillSlot &&
        other.category == category;
  }

  @override
  int get hashCode => Object.hash(step, fillSlot, category);

  @override
  String toString() =>
      'FinishPostMutationTarget(step: $step, fillSlot: $fillSlot, category: $category)';
}

/// Map finish category to set type (identity for armor/weapon/mod).
SetType finishCategoryToSetType(FinishCategory category) {
  switch (category) {
    case FinishCategory.armor:
      return SetType.armor;
    case FinishCategory.weapon:
      return SetType.weapon;
    case FinishCategory.mod:
      return SetType.mod;
  }
}

/// First empty required slot from a gap's emptySlots list order.
String? firstEmptyRequiredSlot(FinishGap gap) {
  if (gap.emptySlots.isEmpty) return null;
  return gap.emptySlots.first;
}

/// Live covering armor set should open optimizer workspace (product 031).
bool shouldOpenArmorOptimize(FinishGap? gap) {
  return gap != null &&
      gap.category == FinishCategory.armor &&
      gap.coveringSetId != null &&
      gap.coveringMode == AttachmentMode.live &&
      gap.status != FinishGapStatus.satisfied;
}

/// Input for [resolvePostMutationStep].
class ResolvePostMutationStepInput {
  const ResolvePostMutationStepInput({
    this.gap,
    this.autoEnterFill = true,
    this.preferArmorOptimize = true,
  });

  final FinishGap? gap;
  final bool autoEnterFill;

  /// When true (default), armor live covering opens optimizer instead of fill.
  final bool preferArmorOptimize;
}

/// Resolve walkthrough step after a finish mutation.
FinishPostMutationTarget resolvePostMutationStep(
  ResolvePostMutationStepInput input,
) {
  final gap = input.gap;
  if (gap == null || gap.status == FinishGapStatus.satisfied) {
    return const FinishPostMutationTarget(
      step: FinishWalkthroughStep.overview,
      fillSlot: null,
      category: null,
    );
  }

  if (input.preferArmorOptimize && shouldOpenArmorOptimize(gap)) {
    return const FinishPostMutationTarget(
      step: FinishWalkthroughStep.armorOptimize,
      fillSlot: null,
      category: FinishCategory.armor,
    );
  }

  if (gap.status == FinishGapStatus.needsFill) {
    if (gap.coveringMode == AttachmentMode.snapshot) {
      return FinishPostMutationTarget(
        step: FinishWalkthroughStep.category,
        fillSlot: null,
        category: gap.category,
      );
    }
    if (gap.coveringMode == AttachmentMode.live && gap.coveringSetId != null) {
      final slot = firstEmptyRequiredSlot(gap);
      if (slot != null && input.autoEnterFill) {
        return FinishPostMutationTarget(
          step: FinishWalkthroughStep.fill,
          fillSlot: slot,
          category: gap.category,
        );
      }
      return FinishPostMutationTarget(
        step: FinishWalkthroughStep.category,
        fillSlot: null,
        category: gap.category,
      );
    }
    return FinishPostMutationTarget(
      step: FinishWalkthroughStep.category,
      fillSlot: null,
      category: gap.category,
    );
  }

  return FinishPostMutationTarget(
    step: FinishWalkthroughStep.category,
    fillSlot: null,
    category: gap.category,
  );
}

/// Whether finish create/capture actions should show for this status.
bool showFinishCreateActions(FinishGapStatus status) {
  return status == FinishGapStatus.needsSet ||
      status == FinishGapStatus.captureAvailable;
}
